text/0000-native-link-modifiers.md

```diff
@@ -0,0 +1,232 @@
+without introducing new linking kinds.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This is an advanced feature not expected to be used commonly,
+see the reference-level explanation if you know that you need some of these modifiers.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Existing syntax of linking attributes and options
+
+- Attributes: `#[link(name = "string", kind = "string", cfg(predicate))]`
+(some components are optional.)
+- Command line options: `-l kind=name:rename` (some components are optional).
+
+## Proposed extensions to the syntax
+
+- Attributes: `#[link(/* same */, modifiers = "+foo,-bar")]`.
+- Command line options: `-l kind:+foo,-bar=name:rename`.
+
+The modifiers are boolean and applied only to the single library specified with `name`. \
+`+` means enable, `-` means disable, multiple options can be separated by commas,
+the last boolean value specified for the given modifier wins. \
+The notation is borrowed from
+[target features](https://doc.rust-lang.org/rustc/codegen-options/index.html#target-feature)
+in general and should have the same semantics.
+
+If the `:rename` component is specified on the command line, then in addition to the name
+and linking kind the modifiers will be updated as well (using concatenation).
+
+## Specific modifiers
+
+### `bundle`
+
+Only compatible with the `static` linking kind.
+
+`+bundle` means objects from the static library are bundled into the produced crate
+(a rlib, for example) and are used from this crate later during linking of the final binary.
+
+`-bundle` means the static library is included into the produced rlib "by name"
+and object files from it are included only during linking of the final binary,
+the file search by that name is also performed during final linking.
+
+This modifier is supposed to supersede the `static-nobundle` linking kind defined by
+[RFC 1717](https://github.com/rust-lang/rfcs/pull/1717).
+
+The default for this modifier is currently `+bundle`,
+but it could be changed later on some future edition boundary.
+
+### `verbatim`
+
+`+verbatim` means that `rustc` won't add any target-specified library prefixes or suffixes
+(like `lib` or `.a`) to the library name.
+
+For `ld`-like linkers `rustc` will use the `-l:filename` syntax (note the colon)
+when passing the library, so the linker won't add any prefixes or suffixes as well. \
+See [`-l namespec`](https://sourceware.org/binutils/docs/ld/Options.html) in `ld` documentation
+for more details.
+
+The default for this modifier is `-verbatim`.
+
+This RFC changes the behavior of `raw-dylib` linking kind specified by
+[RFC 2627](https://github.com/rust-lang/rfcs/pull/2627).
+The `.dll` suffix (or other target-specified suffixes for other targets)
+is now added automatically. \
+If your DLL doesn't have the `.dll` suffix, it can be specified with `+verbatim`.
+
+### `whole-archive`
```

Comment:
I feel like this name is a bit weird, because really it seems to be talking more about LTO. When reading "whole archive" my first thought is "why wouldn't the whole library be linked?"

Reply:
The logic behind both `whole-archive` and `as-needed` is to simply mirror the corresponding linker options.
Generally, if someone wants the "whole-archive" linking, then they probably want it to have the same name "whole-archive" for searchability.
On the other hand, if someone doesn't know what "whole-archive" means, then they probably don't need this option.

Reply:
I'm personally viewing this from the perspective of someone looking at existing code to understand how it works. Documentation will include the options that they translate to for searchability, but I'm not sure someone would need to search for this in code itself.